docs/architecture/adr-031-msg-service.md

```diff
@@ -0,0 +1,222 @@
+# ADR 031: Protobuf Msg Services
+
+## Changelog
+
+- 2020-10-05: Initial Draft
+
+## Status
+
+Proposed
+
+## Context
+
+In early conversations [it was proposed](https://docs.google.com/document/d/1eEgYgvgZqLE45vETjhwIw4VOqK-5hwQtZtjVbiXnIGc/edit)
+that `Msg` return types be captured using a protobuf extension field, ex:
+
+```protobuf
+package cosmos.gov;
+
+message MsgSubmitProposal
+	option (cosmos_proto.msg_return) = “uint64”;
+	string delegator_address = 1;
+	string validator_address = 2;
+	repeated sdk.Coin amount = 3;
+}
+```
+
+This was never adopted, however, and we currently don’t have a mechanism for specifying
+the return type of a `Msg`.
+
+Having a well-specified return value for `Msg`s would improve client UX. For instance,
+in `x/gov`,  `MsgSubmitProposal` returns the proposal ID as a big-endian `uint64`.
+This isn’t really documented anywhere and clients would need to know the internals
+of the SDK to parse that value and return it to users.
+
+Also, there may be cases where we want to use these return values programatically.
+For instance, https://github.com/cosmos/cosmos-sdk/issues/7093 proposes a method for
+doing inter-module Ocaps using the `Msg` router. A well-defined return type would
+improve the developer UX for this approach.
+
+## Decision
+
+We decide to add support for using protobuf `service` definitions for `Msg`s.
+
+A `Msg` `service` definition would be written as follows:
+
+```proto
+package cosmos.gov;
+
+service Msg {
```

Comment:
I think this is a great, standardized way to capture message and response.

I think it is even more useful for queries (where it is already used in the Cosmos SDK). This mapping of query -> request is something I struggle with in API designs for years.

Probably good for each module to define 2 services - one for messages (state changing) and another for queries (immutable)

Reply:
Yeah, that's the idea exactly 👍